src/Psl/Type/Internal/MutableMapType.php

```diff
@@ -0,0 +1,154 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Type\Internal;
+
+use Psl\Collection;
+use Psl\Iter;
+use Psl\Str;
+use Psl\Type;
+use Psl\Type\Exception\AssertException;
+use Psl\Type\Exception\CoercionException;
+
+/**
+ * @template Tk of array-key
+ * @template Tv
+ *
+ * @extends Type\Type<Collection\MutableMapInterface<Tk, Tv>>
+ *
+ * @internal
+ */
+final class MutableMapType extends Type\Type
+{
+    /**
+     * @psalm-var Type\Type<Tk>
+     */
+    private Type\Type $key_type;
+
+    /**
+     * @psalm-var Type\Type<Tv>
+     */
+    private Type\Type $value_type;
+
+    /**
+     * @psalm-param Type\Type<Tk> $key_type
+     * @psalm-param Type\Type<Tv> $value_type
+     */
+    public function __construct(
+        Type\Type $key_type,
+        Type\Type $value_type
+    ) {
+        $this->key_type = $key_type;
+        $this->value_type = $value_type;
+    }
+
+    /**
+     * @psalm-param mixed $value
+     *
+     * @psalm-return Collection\MutableMapInterface<Tk, Tv>
+     *
+     * @throws CoercionException
+     */
+    public function coerce($value): Collection\MutableMapInterface
+    {
+        if (Type\is_iterable($value)) {
+            $key_trace = $this->getTrace()->withFrame(
+                Str\format('%s<%s, _>', Collection\MutableMapInterface::class, $this->key_type->toString())
+            );
+
+            $value_trace = $this->getTrace()->withFrame(
+                Str\format('%s<_, %s>', Collection\MutableMapInterface::class, $this->value_type->toString())
+            );
+
+            /** @psalm-var Type\Type<Tk> $key_type */
+            $key_type = $this->key_type->withTrace($key_trace);
+            /** @psalm-var Type\Type<Tv> $value_type */
+            $value_type = $this->value_type->withTrace($value_trace);
+
+            /**
+             * @psalm-var list<array{0: Tk, 1: Tv}> $entries
+             */
+            $entries = [];
+            /**
+             * @psalm-var mixed $k
+             * @psalm-var mixed $v
+             */
+            foreach ($value as $k => $v) {
+                /** @psalm-var Tk $k */
+                $k = $key_type->coerce($k);
+                /** @psalm-var Tv $v */
+                $v = $value_type->coerce($v);
+
+                $entries[] = [$k, $v];
+            }
+
+            /** @psalm-var Iter\Iterator<Tk, Tv> $iterator */
+            $iterator = Iter\from_entries($entries);
```

Comment:
Would it make sense not to use entries but build the map up from the foreach directly? This would allow for pure / immutable type classes (since `from_entries($iterable)` is not pure)

(same goes for the other usages of entries)

Reply:
it woudln't make any differencce here, since coerce accepts any iterable, and does a foreach on it, meaning it can never be pure. as that iterable we are doing foreach on isn't really immutable.